ledger/src/blockstore.rs

```diff
@@ -1896,6 +1901,121 @@ impl Blockstore {
         None
     }
 
+    /// Parse UpdateParent from data shred payload
+    fn parse_update_parent_from_data_payload(data: &[u8]) -> Option<(Slot, Hash)> {
+        // Try to deserialize as BlockComponent
+        let component = BlockComponent::from_bytes(data).ok()?;
+
+        // If we were able to parse the component, it must be a block marker
+        // NOTE: don't panic if we can't parse the component, since it's possible that we have all
+        // zeroes for the bytes, which shouldn't fail.
+        let marker = component.as_versioned_block_marker()?;
+
+        // Check if it's a BlockMarker with UpdateParent
+        match marker {
+            VersionedBlockMarker::V2(BlockMarkerV2::UpdateParent(versioned_update)) => {
+                // Extract the UpdateParentV1 from the versioned wrapper
+                match versioned_update {
+                    VersionedUpdateParent::V1(update) | VersionedUpdateParent::Current(update) => {
+                        Some((update.new_parent_slot, update.new_parent_block_id))
+                    }
+                }
+            }
+            _ => None,
+        }
+    }
+
+    fn check_for_update_parent(
+        &self,
+        current_shred: &Shred,
+        slot: Slot,
+        location: BlockLocation,
+        write_batch: &mut WriteBatch,
+    ) {
+        // Check if UpdateParentMeta already exists and exit early if duplicate found
+        if let Ok(Some(mut existing_meta)) = self.update_parent_meta_cf.get((slot, location)) {
+            if !existing_meta.duplicate_observed {
+                // Mark as duplicate and update
+                existing_meta.duplicate_observed = true;
+                let _ = self.update_parent_meta_cf.put_in_batch(
+                    write_batch,
+                    (slot, location),
+                    &existing_meta,
+                );
+            }
+            // Already have UpdateParent for this slot/location, no need to continue
+            return;
+        }
+
+        let current_index = current_shred.index();
+        let fec_set_index = current_shred.fec_set_index();
+        let data_complete = current_shred.data_complete();
+
+        // Determine which shred to check for UpdateParent and which FEC set it belongs to
+        let (shred_bytes, target_fec_set) = if data_complete {
+            // Case (a): Current shred has DATA_COMPLETE=true (end of FEC set)
+            // Check the 0th shred in the NEXT FEC set for UpdateParent
+            let next_fec_set_index = fec_set_index + 1;
```

Comment:
The fec set index is actually the first shred index, so we want + 32 here. (annoying I know but it's due to allowing variable sized sets)